### ISSUE TITLE
Out-of-order frames emission

### ISSUE THREAD

**Issue by thaytan** [cisco/openh264#3122]
I've been investigating more of files from the H.264 conformance set that don't match on decode. There are a few cases that fail because frames are not emitted in the right order.

2 good examples are the HCBP1_HHI_A.264 and HCBP2_HHI_A.264 files because the frames otherwise decode pixel-perfect, they're just emitted out of order.

For example, for HCBP1_HHI_A.264 the frames are output in this order:
0, 16, 8, 4, 12, 2, 6, 10, 14, 1, 3, 5, 7, 9, 11, 13, 15....

The JM reference decoder and ffmpeg both output in the correct display order.

**Comment by thaytan** [comment]
I've been looking at other conformance streams where openh264 doesn't produce the right pixels, and not gotten to figuring out what's wrong with the buffer reordering. @xiaotiansf - if you have any spare cycles and want to look at these or other files, I'll keep looking at the incorrect decoding

**Comment by xiaotiansf** [comment]
@thaytan Sure and if you can make the test assets available to me.

**Comment by thaytan** [comment]
The test files are part of the AVC conformance testsuite in https://www.itu.int/net/itu-t/sigdb/spevideo/VideoForm-s.aspx?val=102002641 - the ITU-T_H.264.1(2016-02)_AVCv1_bitstreams.zip files

**Comment by xiaotiansf** [comment]
@thaytan I found that HCBP1_HHI_A.264 and HCBP2_HHI_A.264 contain no B-frames. There is no reordering operation involved. This is quite odd.

**Comment by thaytan** [comment]
You're right. I hadn't noticed that those were baseline streams. I am looking at various files from the conformance set where decoding doesn't match the reference decoder, and picked those two. I don't know how the frames are being emitted out of order there. It's not through the ReorderPicturesInDisplay path.

There are other files in the conformance set that also have reordering issues but most of those also have subtle (or not so subtle) decode errors to unravel, so I didn't suggest them.

CABA3_Sony_C.jsv is one of those others - it is CABAC Main profile with B frames, and outputs out-of-order.

**Comment by huili2** [comment]
For this out-of-order issue, it is a well-known issue as openh264 decoder was first designed for low-delay immediate output purpose. We also have several such kinds of data that differs from standard reference decoder. Till now, we've no plan to modify it.

Thanks again for your input.

**Comment by thaytan** [comment]
Aha! Thanks for the clarification. I assume if we figure it out and fix it you won't mind? It would be good to be sure it still meets the low-delay use case properly - are those covered by the unit tests?

**Comment by huili2** [comment]
Sure if you fix it! Currently there is no UT covering this case.

**Comment by zhutq** [comment]
Hi, I'm facing the same issue. My use case needs the frames exactly in the correct order.
Any updates on this? @thaytan 
Thanks a lot

**Comment by akallabeth** [comment]
might this be related to https://github.com/cisco/openh264/pull/3319 ?

**Comment by xiaotiansf** [comment]
@zhutq Can you provide a sample asset which results in wrong picture order? Thanks.
